### PR TITLE
Fix MKL on mac

### LINKED ISSUE #10685: Darwin support for MKL build / configure script
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Mac OS X 10.12.4, Darwin
- **TensorFlow installed from (source or binary)**:
Source, git tag 1.2.0 RC 2
- **TensorFlow version (use command below)**:
1.2 RC 2
- **Bazel version (if compiling from source)**:
bazel version
..............
Build label: 0.4.3-homebrew
Build target: bazel-out/local-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Thu Dec 22 15:20:15 2016 (1482420015)
Build timestamp: 1482420015
Build timestamp as int: 1482420015
- **CUDA/cuDNN version**:
N/A
- **GPU model and memory**:
N/A
- **Exact command to reproduce**:
./configure 
Please specify the location of python. [Default is /usr/bin/python]: 
Found possible Python library paths:
  /Library/Python/2.7/site-packages
Please input the desired Python library path to use.  Default is [/Library/Python/2.7/site-packages]

Using python library path: /Library/Python/2.7/site-packages
Do you wish to build TensorFlow with MKL support? [y/N] y
MKL support will be enabled for TensorFlow
Do you wish to download MKL LIB from the web? [Y/n] Y
Darwin is unsupported yet

You can collect some of this information using our environment capture script:
https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh
N/A
You can obtain the TensorFlow version with

python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"

### Describe the problem
Darwin / Mac OS X support for Intel's optional MKL optimizations. MKL appears to have darwin support, so I am unsure why it is not supported in Tensorflow. I imagine this is on a to-do, but a publicly tracked feature request might be helpful. Thank you.

### Source code / logs
N/A

### PR DESCRIPTION
**Pull request by anton-matosov**
Fixes #10685
Build can be created with:
```
env TF_MKL_ROOT=/Users/antonmatosov/Downloads/mklml_mac_2018.0.1.20171227 bazel build --config=mkl --config=opt //tensorflow/tools/pip_package:build_pip_package
```

### CODE REVIEW

**Comment by anton-matosov** [comment]
After re-basing on top of latest master there are several new build incompatibilities. Working on them now

**Comment by anton-matosov** [comment]
I have sorted out all build issues. PR is ready for review.
This branch doesn't cover automatic download and bundling of the mkl for mac. I am working on this ATM. I will submit this in a separate PR.

File: tensorflow/core/util/mkl_util.h [hunk 1]
```diff
@@ -1579,10 +1579,10 @@ class MklDnnData {
   }
 
   /// Set function for data buffer of user memory primitive.
-  inline void* SetUsrMemDataHandle(void* data_buffer) {
+  inline void SetUsrMemDataHandle(void* data_buffer) {
     CHECK_NOTNULL(user_memory_);
     CHECK_NOTNULL(data_buffer);
-    return user_memory_->set_data_handle(data_buffer);
+    user_memory_->set_data_handle(data_buffer);
```
Comment (claynerobison) [inline tensorflow/core/util/mkl_util.h]: @nhasabni what are the implications here?

File: tensorflow/tensorflow.bzl [hunk 2]
```diff
@@ -202,7 +202,7 @@ def tf_copts(android_optimization_level_override="-O2", is_external=False):
           "-ftemplate-depth=900"])
       + if_cuda(["-DGOOGLE_CUDA=1"])
       + if_tensorrt(["-DGOOGLE_TENSORRT=1"])
-      + if_mkl(["-DINTEL_MKL=1", "-DEIGEN_USE_VML", "-fopenmp",])
+      + if_mkl(["-DINTEL_MKL=1", "-DEIGEN_USE_VML"])
```
Comment (claynerobison) [inline tensorflow/tensorflow.bzl]: @mahmoud-abuzaina How will this effect Linux builds?

File: tensorflow/core/util/mkl_util.h (hunk 1 above)
Comment (nhasabni) [reply tensorflow/core/util/mkl_util.h]: It was a bug that is fixed here.

File: tensorflow/tensorflow.bzl (hunk 2 above)
Comment (nhasabni) [reply tensorflow/tensorflow.bzl]: We have some op kernels (e.g., Slice) optimized using OpenMP. This change will disable such optimizations unless `-fopenmp` gets passed some other way.

Comment (anton-matosov) [reply tensorflow/tensorflow.bzl]: Thanks for feedback @claynerobison and @nhasabni.
Adding `-fopenmp` under double condition of if_mkl and if_linux appears to be tricky as bazel doesn't support nested select https://github.com/bazelbuild/bazel/issues/1623
The only working solution I have found so far is to update `set_cc_opt_flags` in `configure.py` with these lines
```
  if is_linux():
    write_to_bazelrc("build:mkl --copt=-fopenmp")
    write_to_bazelrc("build:mkl --host_copt=-fopenmp")
```
Please let me know if this is acceptable solution or suggest any other options.

P.S.
I have never used/modified bazel before, so suggestions are very welcome.

**Comment by claynerobison** [comment]
@anton-matosov This PR and 17396 break the MKL Linux builds. Can you test on Linux? Here's the error that we're getting:
 
> bazel build --config=mkl --copt=-mavx2 --copt=-mfma --copt=-march=broadwell --copt=-O3 --copt=-L/opt/tensorflow/gcc6.3/lib64 -s -c opt //tensorflow/tools/pip_package:build_pip_package
> ++ grep 'ERROR: ' ../bazel_build.log
> ++ wc -l
> + '[' 3 = 0 ']'
> + RESULT=FAILURE
> + ERROR=1
> + grep 'ERROR: ' ../bazel_build.log
> ERROR: missing input file '//third_party/mkl:MKL_LICENSE'
> ERROR: /dataset/cje/AIPG_TF/workspace/private-tensorflow-pr-mkldnn/private-tensorflow/tensorflow/tools/pip_package/BUILD:133:1: //tensorflow/tools/pip_package:build_pip_package: missing input file '//third_party/mkl:MKL_LICENSE'
> ERROR: /dataset/cje/AIPG_TF/workspace/private-tensorflow-pr-mkldnn/private-tensorflow/tensorflow/tools/pip_package/BUILD:133:1 1 input file(s) do not exist
> + tail -5 ../bazel_build.log
> Target //tensorflow/tools/pip_package:build_pip_package failed to build
> Use --verbose_failures to see the command lines of failed build steps.
> ERROR: /dataset/cje/AIPG_TF/workspace/private-tensorflow-pr-mkldnn/private-tensorflow/tensorflow/tools/pip_package/BUILD:133:1 1 input file(s) do not exist
> INFO: Elapsed time: 16.226s, Critical Path: 0.27s
> FAILED: Build did NOT complete successfully
> + echo 'RESULT: FAILURE'
> RESULT: FAILURE

**Comment by nhasabni** [comment]
@anton-matosov Renaming of mkl:LICENSE to mkl:MKL_LICENSE in 17396 (https://github.com/tensorflow/tensorflow/pull/17396/files#diff-a36e13dd6539354205d68ce60ea57ea5) seems to have broken the build.

**Comment by claynerobison** [comment]
@opencici2006 can you take a look?

**Comment by anton-matosov** [comment]
#17396 is marked as WIP am still sorting out some issues.
I am now testing build on linux to make sure there are no regressions.
Additionally, it seems, that with existing bugs in bazel it will not be possible to create Mac build without installing MKL dylibs in the library search path. Details are in #17396

**Comment by anton-matosov** [comment]
@claynerobison @nhasabni I have verified MKL build on linux with latest #17396 and it passed without errors:
```
Target //tensorflow/tools/pip_package:build_pip_package up-to-date:
  bazel-bin/tensorflow/tools/pip_package/build_pip_package
INFO: Elapsed time: 1693.973s, Critical Path: 60.63s
INFO: Build completed successfully, 4388 total actions
```
Build command
```
bazel build --config=mkl --config=opt //tensorflow/tools/pip_package:build_pip_package
```
Also tried with the one from your log
```
bazel build --config=mkl --copt=-mavx2 --copt=-mfma --copt=-march=broadwell --copt=-O3 --copt=-L/opt/tensorflow/gcc6.3/lib64 -s -c opt //tensorflow/tools/pip_package:build_pip_package
```
The rename issues was fixed 4 days ago in https://github.com/tensorflow/tensorflow/pull/17396/commits/65e8bd064836156cc699507eb1797d3027435cdb

File: tensorflow/tensorflow.bzl (hunk 2 above)
Comment (gunan) [reply tensorflow/tensorflow.bzl]: One possible way is creating `if_mkl_linux` and `if_mkl_macos` macros.
I have doodled around these a while ago here:
https://github.com/gunan/tensorflow/commit/b4d39c5226e8bbafe084b1dfc35f4644f2d906c2
Feel free to use the changes from there.

Comment (anton-matosov) [reply tensorflow/tensorflow.bzl]: Thanks! Let me try it out.

Comment (anton-matosov) [reply tensorflow/tensorflow.bzl]: That worked! Thanks @gunan! Now I know how to use `config_setting` =)

**Comment by opencici2006** [comment]
Hi @anton-matosov ,

We have encountered problem when building TF with MKL on mac as is written in [https://github.com/bazelbuild/bazel/issues/4794](url) .

After communications in the same page, it is recommended to try the newest bazel non -  released, so we think about compiling bazel from source. This requires the installation of a released version, and problems still occur when installing the released bazel v 0.11.1 with a direct `bash ./compile.sh` . 

Maybe you would be interested to try building bazel from source code on mac, and generously share with us a built binary?

**Comment by anton-matosov** [comment]
Hi @opencici2006,
Thanks for trying this branch out. I believe you shared wrong link, the one you shared is about a broken URL in bazel's readme file.

Building bazel is a bit cumbersome and will require tweaking build arguments later (e.g. manually built version doesn't infer cpu on mac for some reason, so you need to pass `--cpu=darwin` for build to pass). So I discourage this way.

As discussed in #17396 I have used rc1 build version made available by bazel team. So this is better approach than building yourself.

And the simplest way to bypass bazel linker issues is to simply install MKL dylibs in your system to `/usr/local/lib`. You can find detailed instructions in my original guide on building MKL here https://github.com/anton-matosov/tensorflow-wheels/tree/master/Tensorflow-Wheels/MacOS/MKL

**Comment by anton-matosov** [comment]
This PR is superseded by #17396

**Comment by opencici2006** [comment]
Hi @anton-matosov 
Thanks for your reply, I have updated the issue link.
https://github.com/bazelbuild/bazel/issues/4794

You could see our changes similiary as your PR. We will have a try following your instructions guide to bypass bazel linker issue.

Thanks a lot.

**Comment by anton-matosov** [comment]
You are very welcome. Bazel 0.12 is finally out, so no more manual workarounds needed.

**Pull request closed without merging**